clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+
+  static void setKernelImplicitAttrs(ASTContext &Context, FunctionDecl *FD,
+                                     StringRef Name) {
+    // Set implict attributes.
+    FD->addAttr(OpenCLKernelAttr::CreateImplicit(Context));
+    FD->addAttr(AsmLabelAttr::CreateImplicit(Context, Name));
+    FD->addAttr(ArtificialAttr::CreateImplicit(Context));
+  }
+
+  static FunctionDecl *initKernelObj(ASTContext &Ctx, StringRef Name,
+                                     SourceLocation Loc, bool IsInline) {
+    // Create this with no prototype, and we can fix this up after we've seen
+    // all the params.
+    FunctionProtoType::ExtProtoInfo Info(CC_OpenCLKernel);
+    QualType FuncType = Ctx.getFunctionType(Ctx.VoidTy, {}, Info);
+
+    FunctionDecl *FD = FunctionDecl::Create(
+        Ctx, Ctx.getTranslationUnitDecl(), Loc, Loc, &Ctx.Idents.get(Name),
+        FuncType, Ctx.getTrivialTypeSourceInfo(Ctx.VoidTy), SC_None);
+    FD->setImplicitlyInline(IsInline);
+    setKernelImplicitAttrs(Ctx, FD, Name);
+
+    // Add kernel to translation unit to see it in AST-dump.
+    Ctx.getTranslationUnitDecl()->addDecl(FD);
+    return FD;
+  }
+
+public:
+  SyclKernelDeclCreator(Sema &S, SyclKernelFieldChecker &ArgChecker,
+                        StringRef Name, SourceLocation Loc, bool IsInline)
+      : SyclKernelDeclBase(
+            S, initKernelObj(S.getASTContext(), Name, Loc, IsInline), ArgTys,
+            Params),
+        ArgChecker(ArgChecker), FuncContext(SemaRef, KernelObj) {}
+
+  ~SyclKernelDeclCreator() {
+    KernelObj->setParams(Params);
+
+    if (ArgChecker.isValid())
+      SemaRef.addSyclDeviceDecl(KernelObj);
+  }
+
+  void handleSyclSamplerType(const FieldDecl *FD, QualType ArgTy) final {
+    handleSpecialType(FD, ArgTy);
+  }
+
+  void handlePointerType(const FieldDecl *FD, QualType ArgTy) final {
+    // TODO: Can we document what the heck this is doing?!
+    QualType PointeeTy = ArgTy->getPointeeType();
+    Qualifiers Quals = PointeeTy.getQualifiers();
+    Quals.setAddressSpace(LangAS::opencl_global);
+    PointeeTy = SemaRef.getASTContext().getQualifiedType(
+        PointeeTy.getUnqualifiedType(), Quals);
+    QualType ModTy = SemaRef.getASTContext().getPointerType(PointeeTy);
+    addParam(FD, ModTy);
+  }
+  void handleScalarType(const FieldDecl *FD, QualType ArgTy) final {
+    addParam(FD, ArgTy);
+  }
+
+  // This is implemented here because this is the only case where the recurse
+  // object is required.  The base type is pretty cheap, so we might opt
+  // to just always create it (the way this one is implemented) and just put
+  // this implementation in the base.
+  void handleStructType(const FieldDecl *FD, QualType ArgTy) final {
+    addParam(FD, ArgTy);
+
+    // Create descriptors for each accessor field in the class or struct
+    const auto *Wrapper = ArgTy->getAsCXXRecordDecl();
+    SyclKernelDeclBase Recurse(SemaRef, KernelObj, ArgTys, Params);
+    VisitKernelFields(Wrapper->fields(), Recurse);
```

Comment:
The `Wrapper` here is not a SYCL kernel object. Seeing Visit**Kernel**Fields seems strange here. Maybe we probably need to rename `VisitKernelFields` to something like `visitRecordFields`.

Reply:
Sure.